### ISSUE TITLE
Closures module

### ISSUE THREAD

**Issue by ntarocco** [inveniosoftware/invenio-app-ils#843]
# TODO

* create user stories, similarly to what done [here](https://github.com/inveniosoftware/invenio-app-ils/issues/637)
* create tasks

## User stories

https://codimd.web.cern.ch/tXelz_qxQCyDs3OoRJTpNA

## Use cases from brainstorming

* Find the right name for this feature in the English dictionary and use it consistently. 
* Implement CRUD for:
  * opening/closing days and hours (hours just for display, not affecting circulation)
  * exceptions (allow single days or a period of days of opening or closing)
  * update the Location schema and edit form to store such information
* On the librarian changes opening/exceptions, a task should update all existing loans:
  * for active loans, change `end_date` or trigger an extension with a specific `end date`. The extension has the advantage of already sending an e-mail to the patron, but extensions have constraints to be checked (max n. of extension, book requested?)
  * for pending loans, change the `request_expire_date`
  * for ILLs (`item.pid_type: illbid`), it should be the same as normal loans
* Patrons should receive an email notification when the end date has been changed (already done if extension, please verify)
* Run a cron job to update the opening/closing days once a day when opening/exceptions are changed, so that eventually in case of mistake, the librarian can correct the opening/exceptions dates multiple times, without trigger immediate changes to loans.
* Create a dedicated front site page to render the opening/closing days and exceptions for each Location, fetched from the db. Nice to have: if logged in, show patron's Location as first.
* Opening/closing days are expsoed in the `/api/locations` endpoint, just change the permissions to be any user (currently only admins probably).
* The calendar React widget should be aware of the opening/closing days in the following places:
  * In the front site document page (Do you require it before a certain date?) Display based on patron location
  * backoffice document request a loan for document, display based on patron location
  * backoffice borrowing request in requested state when clicking new loan button, display based on patron location
* Endpoint cache: the REST endpoint is actually never cached on purpose, so limit the number of request from the React app if possible (fetch only once, maybe with a timeout?)

**Nice to have:**
* When changing opening/exceptions, the librarian should see a text mentioning how many loans are affected by the change before confirming. The link should clickable to open a new tab with a search on loans. e.g. `Changing these dates will affect _3 requests_ and _10 active loans_`

## Implementation (sketch)

https://codimd.web.cern.ch/CS7l855FTli6ilx-q-uhkA

## Extra

See: https://ils.test.rero.ch/

RERO-ILS implemented such a module, here their explanation:
```
In the json of the resource library we have defined the opening_hours and exception_dates -> https://github.com/rero/rero-ils/blob/78e62f040d809450eac056139cebcffad3caa4cf/rero_ils/modules/libraries/jsonschemas/libraries/library-v0.0.1.json#L62

Thanks to these, we are able to check if a library is open or not, and to know the next open day for a long loan period -> https://github.com/rero/rero-ils/blob/78e62f040d809450eac056139cebcffad3caa4cf/rero_ils/modules/libraries/api.py#L205
and
https://github.com/rero/rero-ils/blob/78e62f040d809450eac056139cebcffad3caa4cf/rero_ils/modules/libraries/api.py#L237

Lastly, we uses the library.next_open and library.is_open to pass the right loan.end_date to Invenio-circulation
Two things specific à RERO:
* The decisions are made according to the library of the transaction location and NOT the owning library of the item
* End dates are all pushed to the end of the day i.e. 23:59
```
